h2o-persist-s3/src/main/java/water/persist/PersistS3.java

```diff
@@ -169,8 +150,77 @@ public InputStream open(String path) {
     return s3obj.getObjectContent();
   }
 
-  public static InputStream openStream(Key k, RIStream.ProgressMonitor pmon) throws IOException {
-    return new H2SO3InputStream(k, pmon);
+  @Override
+  public OutputStream create(String path, boolean overwrite) {
+    String[] bk = decodePath(path);
+    final File tmpFile;
+    try {
+      tmpFile = File.createTempFile("h2o-export", ".bin");
+      tmpFile.deleteOnExit();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to create temporary file for S3 object upload", e);
+    }
+    Runnable callback = new PutObjectCallback(getClient(), tmpFile, true, bk[0], bk[1]);
+    try {
+      return new CallbackFileOutputStream(tmpFile, callback);
+    } catch (FileNotFoundException e) {
+      throw new RuntimeException(e); // should never happen
+    }
+  }
+  
+  static class PutObjectCallback implements Runnable {
+    private final AmazonS3 _client;
+    private final File _file;
+    private final boolean _deleteOnDone;
+    private final String _bucketName;
+    private final String _key;
+
+    public PutObjectCallback(AmazonS3 client, File file, boolean deleteOnDone, String bucketName, String key) {
+      _client = client;
+      _file = file;
+      _deleteOnDone = deleteOnDone;
+      _bucketName = bucketName;
+      _key = key;
+    }
+
+    @Override
+    public void run() {
+      try {
+        PutObjectRequest request = new PutObjectRequest(_bucketName, _key, _file);
+        PutObjectResult result = _client.putObject(request);
+        Log.info("Object `" + _key + "` uploaded to bucket `" + _bucketName + "`, ETag=`" + result.getETag() + "`.");
+      } finally {
+        if (_deleteOnDone) {
+          boolean deleted = _file.delete();
+          if (!deleted) {
+            LOG.warn("Temporary file `" + _file.getAbsolutePath() + "` was not deleted. Please delete manually.");
+          }
+        }
+      }
+    }
+  } 
+
+  static class CallbackFileOutputStream extends FileOutputStream {
+    private final Object closeLock = new Object();
+    private volatile boolean closed = false;
```

Comment:
`volatile` is redundant when accessing only from synchronized block

Reply:
good catch, this is just copy&paste from FileOutputStream; they are doing the same thing - I can remove it but I would rather keep it identical

@koniecsveta WDYT?

Reply:
:+1: